0/TransactionBuilder.js

```diff
@@ -44,38 +44,79 @@ class TransactionBuilder {
      * @returns {TransactionBuilder}
      */
     input(fs, amount) {
+
+        if ( this._signatures !== undefined ) {
+            throw new Error("Attempting to add new input, expecting signatures only")
+        }
         //if this is setup as coinbase, prevent additional inputs
         if (Object.keys(this._inputs).find(address => address === constant.COINBASE_ADDRESS_PUBLIC)) throw new Error('Cannot add an additional input to a coinbase transaction');
 
-        if (!fctAddressUtil.isValidPrivateAddress(fs)) throw new Error("Input address must be a valid private Factoid address");
+        //if it isn't a private address and instead a public address then, the fs should be a public key      
+        if (fctAddressUtil.isValidPrivateAddress(fs)) { //first check to see if valid private address
 
-        amount = new BigNumber(amount);
-        if (!amount.isInteger() || amount.isLessThan(1)) throw new Error("Input amount must be a positive nonzero integer");
+            amount = new BigNumber(amount);
+            if (!amount.isInteger() || amount.isLessThan(1)) throw new Error("Input amount must be a positive nonzero integer");
+
+            //check that outputs does not contain this address
+            if (this._outputs[fctAddressUtil.getPublicAddress(fs)]) throw new Error("Input address already occurs in outputs");
+
+            this._keys.push(nacl.keyPair.fromSeed(fctAddressUtil.addressToKey(fs)));
+            this._inputs[fctAddressUtil.getPublicAddress(fs)] = amount;
 
-        //check that outputs does not contain this address
-        if (this._outputs[fctAddressUtil.getPublicAddress(fs)]) throw new Error("Input address already occurs in outputs");
+        } else {
 
-        this._keys.push(nacl.keyPair.fromSeed(fctAddressUtil.addressToKey(fs)));
-        this._inputs[fctAddressUtil.getPublicAddress(fs)] = amount;
+            if ( fctAddressUtil.isValidPublicFctAddress(fs) ) { //check to see if user passed in a public fct address
+              throw new Error("Input address must be either a valid private Factoid address or a Factoid public key");
+            }
+
+            amount = new BigNumber(amount);
+            if (!amount.isInteger() || amount.isLessThan(1)) throw new Error("Input amount must be a positive nonzero integer");
+
+            // we'll have to assume that the fs is really the publicKey if we get this far, ultimately the tx will fail if not
+            let pubkey = fs;
+
+            //this will throw exception if not right length
+            let pubaddr = fctAddressUtil.keyToPublicFctAddress(pubkey);
+
+            //check that outputs does not contain this address
+            if (this._outputs[pubaddr]) throw new Error("Input address already occurs in outputs");
+
+            this._keys.push({publicKey: pubkey});
+            this._inputs[pubaddr] = amount;
+        }
         return this;
     }
 
```

Comment:
Yes, that is true, good catch.  I was thinking from the perspective of the Ledger, which unfortunately can't support coinbase  transactions since it uses the sk keys instead of the idsec/idpub.

Reply:
added support for external signing for coinbase transactions